File: 2018/day07_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Step = tuple[str, list[str]]


def solution(content: list[str]) -> Any:
    names = set(line.split(" ")[1] for line in content) | set(
        line.split(" ")[-3] for line in content
    )
    steps: list[Step] = [(name, parse_requirements(content, name)) for name in names]
    return unwind(steps)


def unwind(steps: list[Step]) -> str:
    output = ""
    copy = steps.copy()
    while copy:
        first, *_ = sorted(step[0] for step in copy if not step[1])
        output += first
        copy = [step for step in copy if step[0] != first]
        copy = [(name, [req for req in reqs if req != first]) for (name, reqs) in copy]

    return output


def parse_requirements(content: list[str], name: str) -> list[str]:
    return [line.split(" ")[1] for line in content if line.split(" ")[-3] == name]


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2018/day07-1.ts
type Step = { name: string; requires: string[] };

function solution(content: string) {
  const lines = content.split("\n");
  const names = new Set(
    lines
      .map((line) => line.split(" ")[1])
      .concat(lines.map((line) => line.split(" ").at(-3)!))
  );
  const steps: Step[] = Array.from(names).map((name) => ({
    name,
    requires: lines
      .map((line) => line.split(" "))
      .filter((items) => items.at(-3) === name)
      .map((items) => items[1]),
  }));
  return unwind(steps);
}

function unwind(steps: Step[]) {
  let copy = steps.slice();
  let output = "";
  while (copy.length) {
    const [first, ..._rest] = copy
      .filter((step) => !step.requires.length)
      .map((step) => step.name)
      .toSorted();
    output += first;
    copy = copy.filter((step) => step.name !== first);
    copy = copy.map(({ name, requires }) => ({
      name,
      requires: requires.filter((req) => req !== first),
    }));
  }
  return output;
}

Deno.readTextFile("day07.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
